types/part_set.go

```diff
@@ -29,7 +31,11 @@ type Part struct {
 // ValidateBasic performs basic validation.
 func (part *Part) ValidateBasic() error {
 	if len(part.Bytes) > int(BlockPartSizeBytes) {
-		return fmt.Errorf("too big: %d bytes, max: %d", len(part.Bytes), BlockPartSizeBytes)
+		return ErrPartTooBig
+	}
+	// All parts except the last one should have the same constant size.
+	if int64(part.Index) < part.Proof.Total-1 && len(part.Bytes) != int(BlockPartSizeBytes) {
```

Comment:
We're still using `part.Proof.Total` here, are 100% sure this total will always be in sync with the total number of parts?

Reply:
Yes, this is a good point. But we don't have the `Total` field in each `Part`, we only have it on the `PartSet`.

Reply:
The point here is that wrong values for `Proof.Total` and `Proof.Index` should make the Merkle-tree verification to fail. But I am not sure whether we test this at this point. Notice that the `Part` tests have nothing on the arrays used to fill the `Part` and the `Part.Proof` bytes.

Reply:
Is there a possibility that `Part` is used without populating `part.Proof`? (I guess no, but double-checking).

Reply:
It should not be possible, as the next check (line 40) would fail. This check, however, is also a valid check: integer fields are >= 0, hashes have the right length.

Reply:
So, status regarding this:
- we can trust `Proof.Index`, as change it causes the proof verification to fail
- we can't trust `Proof.Total`, as it is irrelevant for the proof verification provided that `Proof.Index < Proof.Total`.

Reply:
Ok, now `Proof.Total` should be correct, otherwise the `Part` is not added. Lets check if something breaks due to this additional check.

Reply:
Excellent!